DOC: ACSIncome Documentation (#1005)
fairlearn/fairlearn: docs/user_guide/datasets/acs_income.rst

.. _acsincome-data:

> Retiring Adult: ACSIncome

Reviewer: I believe their paper title was meant to be a pun about "retiring" the UCI adult dataset, so I'm not sure that it makes sense to include that pun as the title here, since it may make readers think it's a dataset about retiring adults..

Author: Fair point. I'll just say ACSIncome here to improve clarity. Thanks!

-------------------------

Introduction
^^^^^^^^^^^^^^^^^

The ACSIncome dataset is one of five datasets created by Ding et al. [1]_ 
as an improved alternative to the popular UCI Adult dataset.
The authors compiled data from the American Community Survey (ACS) Public Use Microdata Sample (PUMS). 
Note that this is a different source than the Annual Social and Economic Supplement (ASEC) 
of the Current Population Survey (CPS) used to construct the original UCI Adult dataset.
Ding et al. [1]_ filtered the data such that ACSIncome only includes individuals above 16 years old 

> who worked at least 1 hour per week in the past year and had an income of at least $100.

Reviewer: A few words about why they filtered it that way could be helpful

Author: I'm actually unsure why these filters were chosen. For example, page 7 of the paper simply states

```
ACSIncome: predict whether an individual’s income is above $50,000, after filtering the ACS PUMS data sample to only include individuals above the age of 16, who reported usual working hours of at least 1 hour per week in the past year, and an income of at least $100.
```

Page 21 in the SI restates the filters, but again does not offer an explanation. In contrast, the ACSPublicCoverage is filtered to individuals under the age of 65, and the authors explicitly state that the purpose was to focus on people who are not eligible for Medicare.

.. _acsincome_dataset_description:

Dataset Description
^^^^^^^^^^^^^^^^^^^
The authors provide data from 2014-2018 for all 50 states and Puerto Rico.

> We uploaded the 2018 data to OpenML, which fairlearn can access.

Reviewer 2: Do we want to eventually support the ACS datasets from the other years (2014-2017) as well? One of the key benefits of these datasets, as listed in the paper by the authors, was to perform temporal evaluation of a fairness criterion and intervention.

Author: Great question. Yes I think we want to support these in the future. My opinion was that the scope of this PR should only add 2018 data for one of the datasets (ACSIncome) as a starting point. Ideas for future PRs include 
1) Adding other datasets from the paper, such as ACSPublicCoverage which @britneyting is working on
2) Adding analysis of how the composition of the new ACSIncome dataset compares to UCI Adult. Some interesting questions could include: Are certain groups better represented in the updated dataset? Are any discrepancies larger or smaller? 
3) Adding data from other years

.. _acsincome-data:
ACSIncome
-------------------------

Introduction
^^^^^^^^^^^^^^^^^

The ACSIncome dataset is one of five datasets created by Ding et al. [1]_ 
as an improved alternative to the popular UCI Adult dataset [2]_.

> Briefly, the UCI Adult dataset is commonly used as a benchmark dataset when comparing

Reviewer 3: I'm not sure what "Briefly" is meant to indicate here tbh. Perhaps it's because English is not my native language (?)

different algorithmic fairness interventions. ACSIncome offers a few improvements,
such as providing more datapoints (1,664,500 vs. 48,842) and more recent data (2018 vs. 1994).

> Further, the binary labels in the UCI Adult dataset indicate whether an indivdual earned 

Reviewer 3: ```suggestion
Further, the binary labels in the UCI Adult dataset indicate whether an individual earned 
```

Author: Thanks for catching another spelling mistake! Odd that my editor didn't alert me. I'll check my settings (I actually just caught another)

more than $50k US dollars in that year. Ding et al. show that the choice of threshold impacts

> the amount of fairness violation, so they allow users to define any threshold.

Reviewer 3: I'd suggest adding at the end: "rather than fixing it at $50k." to be super explicit.

Author: Good addition!

Ding et al. compiled data from the American Community Survey (ACS) Public Use Microdata Sample (PUMS). 
Note that this is a different source than the Annual Social and Economic Supplement (ASEC) 
of the Current Population Survey (CPS) used to construct the original UCI Adult dataset.
Ding et al. filtered the data such that ACSIncome only includes individuals above 16 years old 
who worked at least 1 hour per week in the past year and had an income of at least $100.

.. _acsincome_dataset_description:

Dataset Description
^^^^^^^^^^^^^^^^^^^

> Ding et al. provide data from 2014-2018 for all 50 states and Puerto Rico.

Reviewer 3: I'm undecided on whether it makes sense to explicitly mention that other territories are excluded. What do you think?

Author: In my opinion, the documentation should just describe what is included and not necessarily enumerate everything that is excluded, but I'm open

Reviewer 3: I'm thinking from my perspective as someone who wasn't born in the US. I don't think I knew about anything beyond 50 states, let alone territories like Guam, until I moved to the US. From that mistaken viewpoint it's easy to assume you're querying all of the US and describe it as such. In other words, this nuance is potentially lost on people. Part of why we're documenting this is to avoid missing nuances like that. This is the application context that's so crucial. If we train a model on data from 50 states + PR we can't assume it works elsewhere (not even the territories). [In fact, we shouldn't necessarily assume it "works" anywhere unless we validate it in its real world context.]

So perhaps this makes sense as elaborating a little bit on why it's a slippery slope, and the more information the better.

Author: Thanks for clarifying. I added a sentence of "Note that Puerto Rico is the only US territory included in this dataset." Let me know if we should provide more detail like explicitly naming the territories that are excluded

> We uploaded the 2018 data to OpenML, which fairlearn can access.

Reviewer 3: Anyone can access it 🙂 Also, Fairlearn is capitalized unless we're in code `fairlearn`

Author: Updated this description

         7. Self-employed in own incorporated business, professional practice or farm 
         8. Working without pay in family business or farm 
         9. Unemployed and last worked 5 years ago or earlier or never worked

   *  - SCHL
      - Educational attainment:
         1. No schooling completed
         2. Nursery school or preschool
         3. Kindergarten
         4. Grade 1
         5. Grade 2
         6. Grade 3
         7. Grade 4
         8. Grade 5
         9. Grade 6
         10. Grade 7
         11. Grade 8
         12. Grade 9
         13. Grade 10
         14. Grade 11
         15. Grade 12 (no diploma)
         16. Regular high school diploma
         17. GED or alternative credential
         18. Some college but less than 1 year
         19. 1 or more years of college credit but no degree
         20. Associate's degree
         21. Bachelor's degree
         22. Master's degree
         23. Professional degree beyond a bachelor's degree
         24. Doctorate degree

   *  - MAR
      - Marital status:
         1. Married
         2. Widowed
         3. Divorced
         4. Separated
         5. Never married or under 15 years old

   *  - OCCP

>       - Occupation. There are over 500 categories. Please see data dictionary at `ACS PUMS documentation <https://www.census.gov/programs-surveys/acs/microdata/documentation.2018.html>`_ for the full list of occupation codes.

Reviewer 3: Can you try to keep the line lengths to 79 chars max? There are cases where a link is longer and links don't split well in ReST, so it's best to keep them on a single line, but for everything else it should be possible.

Author: I edited the paragraphs to be < 79 characters. However, this section is a table so adding newlines was creating errors when compiling the documentation

.. _acsincome-data:
Retiring Adult: ACSIncome
-------------------------

Introduction
^^^^^^^^^^^^^^^^^

> The ACSIncome dataset is one of five datasets created by Ding et al. [#1]_ 

Reviewer 4: I see two [2] in the preview:

![image](https://user-images.githubusercontent.com/24683184/148844966-21989cf1-d1d3-4d55-8bbb-c5d213d45aa1.png)

Author: Thanks for pointing this out. I pushed a fix. Once the PR is passing tests and ready to merge, I can squash the temporary fixup commits

Reviewer 3: Don't worry about commits. We're squashing anyway.

Author: Is the plan to squash everything into one commit or do we squash into the original 4 commits?

Reviewer 3: We do that upon merging, no need to worry about it on your end.

> as an improved alternative to the popular UCI Adult dataset.

Reviewer 3: Are you planning on adding a description for UCI Adult as well? Just wondering because this leaves it unclear what the improvements actually are, and what the problems with UCI Adult were in the first place. It could totally be a separate page, of course. You can absolutely borrow heavily from the paper (with citation). No need to reinvent the wheel 😉

Author: I wasn't planning to, at least not in this PR. To be honest, I was a bit confused what the improvements were. It's definitely more data (1.6M rows vs. ~40k), and it's more recent data (2018 vs. 1994). The paper questions the choice of $50k income cutoff, which is why we're providing the raw incomes here. I'd need to do more analysis to understand how the composition of these datasets differ. Let's discuss more as a team about what we think a good dataset for fairness benchmarking means

Reviewer 5: If we do end up adding more information to this page about issues with the UCI Adult dataset (and I agree that even a few paragraphs would provide useful context), it might be good to move the information about the unique values in each variable to an appendix/the end of the page and just include the field names + description in this table; otherwise, the length of the table might be unwieldy.

Reviewer 3: One of the reasons I'm asking: UCI Adult documentation might be the foundation for documentation of these ACS datasets, and perhaps on a separate page.

Re: limitations: I definitely agree on dataset size and recency as well as the choice of cutoff. Others from the paper:
> The choice of the 50,000 dollar threshold is idiosyncratic and potentially limits the external validity of UCI Adult as a benchmark for algorithmic fairness. In 1994, the median USincome was 26,000 dollars, and 50,000 dollars corresponds to the 76th quantile of the income distribution, and the 88th and 89th quantiles of the income distribution for the Black and female populations, respectively. Consequently, almost all of the Black and female instances in the dataset fall below the threshold and models trained on UCI adult tend to have substantially higher accuracies on these subpopulations. For instance, a standard logistic regression model trained on UCI Adult dataset achieves 85% accuracy overall, 91.4% accuracy on the Black instances, and 92.7% on Female instances. This is a rather untypical situation since often machine learning models perform more poorly on historically disadvantaged groups.
The following chart also tells you how much influence this choice of threshold has on the fairness objectives that we're trying to optimize for: 
![image](https://user-images.githubusercontent.com/10245648/149057359-7320374a-5ed7-47ec-a2c1-eea47e8a7d97.png)

More: 
> A notable if obvious limitation of our work is that it is entirely US-centric. A richer dataset ecosystem covering international contexts within the algorithmic fairness community is still lacking. Although empirical work in the Global South is central in other disciplines, there continues to be much need for the North American fairness community to engage with it more strongly

Another limitation IMO is the task definition that doesn't strike me as realistic. After all, why would you want to predict people's income? There's probably a different motivation there. That's probably a larger point of discussion, though. The main ones I wanted to include are the ones mentioned in the paper. I agree that this may be better off in a separate PR, though.

@alliesaizan what do you mean by "unique values in each variable"? Did you mean the feature descriptions? I personally don't feel very strongly about including in the text or at the end, but it's certainly something we can consider as more content is added.

Reviewer 5: @romanlutz my apologies if I was not clear! I mean putting the part of the feature description that reviews what values the variable can take at the end, but agree that we can reevaluate this suggestion when more content has been added.

Reviewer: Agreed with @romanlutz that even a sentence or two of context for what the UCI Adult dataset is and why an improved alternative might be needed will be helpful for readers who may not have that context!

Author: I added a brief summary of UCI Adult and the improvements offered by ACSIncome based on the points made in this thread :)

The authors compiled data from the American Community Survey (ACS) Public Use Microdata Sample (PUMS). 
Note that this is a different source than the Annual Social and Economic Supplement (ASEC) 
of the Current Population Survey (CPS) used to construct the original UCI Adult dataset.
Ding et al. [#0]_ filtered the data such that ACSIncome only includes individuals above 16 years old 
who worked at least 1 hour per week in the past year and had an income of at least $100.

.. _acsincome_dataset_description:

Dataset Description
^^^^^^^^^^^^^^^^^^^

> Data is provided from 2018 for all 50 states and Puerto Rico.

Reviewer 5: ```suggestion
The authors provide 2018 data for all 50 states and Puerto Rico.
```

Reviewer 5: Removing passive voice :)

> There are 1,664,500 rows and 10 features, which are described below:

Reviewer 5: ```suggestion
The dataset contains 1,664,500 rows and 10 features, which are described below:
```

Reviewer 3: Well spotted! If you want to go a step further you can also replace the "are described" with "we describe" or something like that.

Reviewer 6: +1 on having more active voice.

> .. list-table::

Reviewer 5: This table looks pretty long. What do you think about just including the field names + descriptions in the table, and listing the unique values within each variable in an appendix?

   :header-rows: 1
   :widths: 7 30

>    :stub-columns: 1

Reviewer 3: It's a bit hard to imagine what this looks like when built. Can you run the doc build command and paste what this looks like locally for you?

Author: Admittedly, I just copied the table settings from the boston housing file. This is how it looks on my end. Let me know if I should make changes
<img width="800" alt="Screen Shot 2022-01-11 at 4 05 33 PM" src="https://user-images.githubusercontent.com/31116084/149021373-1eecb9ed-2ab8-45ee-b379-375169a69477.png">

Reviewer 3: Looks fine! Thank you

         4. Separated
         5. Never married or under 15 years old

   *  - OCCP
      - Occupation. There are over 500 categories. Please see data dictionary at `ACS PUMS documentation <https://www.census.gov/programs-surveys/acs/microdata/documentation.2018.html>`_ for the full list of occupation codes.

   *  - POBP
      - Place of birth. There are over 200 categories, including the 50 US states and several countries. Please see the data dictionary at `ACS PUMS documentation <https://www.census.gov/programs-surveys/acs/microdata/documentation.2018.html>`_ for the full list.

   *  - RELP
      - Relationship:
         0. Reference person
         1. Husband or wife
         2. Biological son or daughter
         3. Adopted son or daughter
         4. Stepson or stepdaughter
         5. Brother or sister
         6. Father or mother
         7. Grandchild
         8. Parent-in-law
         9. Son-in-law or daughter-in-law
         10. Other relative
         11. Roomer or boarder
         12. Housemate or roommate
         13. Unmarried partner
         14. Foster child
         15. Other nonrelative
         16. Institutionalized group quarters population

>          17. Noninstitutionalized group quarters population

Reviewer 3: What do these actually mean? Any idea? I'm not familiar with these terms

Reviewer 7: According to [disabilitystatistics](https://www.disabilitystatistics.org/glossary.cfm?g_id=266&view=true#nextTerm263), "institutionalized population includes people under formally authorized supervised care," and those people are generally referred to as "inmates" or "patients" and have restricted access to the surrounding community, whereas "the noninstitutionalized population includes all people who live in group quarters other than institutions, such as dormitories, halfway houses, and communes."

Reviewer 5: I believe the Census uses these terms to distinguish group quarters like university dorms from group quarters like prison facilities. Though I agree that a footnote explaining the distinction would be helpful! Here's more background: https://www.census.gov/topics/income-poverty/poverty/guidance/group-quarters.html

Reviewer 3: Couldn't agree more with both of you here! I find this context very useful to understand as well.

   *  - WKHP
      - Usual hours worked per week in the past 12 months. Values are an integer from 1 to 99. Any hours above 99 are rounded down to 99

   *  - SEX
      - Sex code:
         1. Male
         2. Female

   *  - RAC1P
      - Race code
         1. White alone
         2. Black or African American alone
         3. American Indian alone
         4. Alaska Native alone
         5. American Indian and Alaska native tribes specified; or American Indian or Alaska Native, not specified and no other races
         6. Asian alone
         7. Native Hawaiian and Other Pacific Islander alone
         8. Some Other Race alone
         9. Two or More races

The target label is given by PINCP. For generalizability, the integer value is provided.
A threshold can be applied to PINCP to frame this as a binary classification task.

.. list-table::
   :header-rows: 1
   :widths: 7 30
   :stub-columns: 1

   *  - Column name
      - Description

   *  - PINCP

>       - Total income, denoted as an integer ranging from 104 to 1,423,000.

Reviewer 6: Income per ...... ?

Reviewer 6: I was actually thinking weekly vs monthly vs annual, but also good to clarify personal vs household.

Author: Ah thanks for that clarification. I'll add the word annual now!

> An additonal column for the state code is also provided so users to easily filter the data to focus on subsets by state.

Reviewer 5: ```suggestion
An additional state code column allows users to easily filter the data to focus on subsets by state.
```

Reviewer 6: Also 'easily' is splitting an infinitive (not the worst of sins, but I don't think it adds anything here), and is a bit redundant.

Author: Based on @romanlutz comment above, I'm changing the function so that the state code is no longer returned. So I'll remove this part from the documentation

.. _acsincome-data:
ACSIncome
---------

Introduction
^^^^^^^^^^^^

The ACSIncome dataset is one of five datasets created by Ding et al. [1]_ 
as an improved alternative to the popular UCI Adult dataset [2]_.
Briefly, the UCI Adult dataset is commonly used as a benchmark dataset 
when comparing different algorithmic fairness interventions. ACSIncome offers 
a few improvements, such as providing more datapoints (1,664,500 vs. 48,842) 
and more recent data (2018 vs. 1994). Further, the binary labels in the UCI 
Adult dataset indicate whether an individual earned more than $50k US dollars 
in that year. Ding et al. show that the choice of threshold impacts the 

> amount of fairness violation, so they allow users to define any threshold 

Author: Thanks for the suggestion. I updated this wording to use "amount of disparity"

.. _acsincome_data:

> ACSIncome

Reviewer 8: [I haven't gone through the other PR yet so this may be irrelevant]

It seems to me that a user guide item should also provide guidance on when one should (not) use this data set (or at least refer to other resources that cover that).

Author: We agree. However, we were thinking that the scope of this PR is just to report the facts of what this dataset is. Any recommendations/interpretations would be a good candidate for a future PR since it's more of an open discussion that definitely requires further analysis to justify any recommendations, especially since UCI Adult (and this updated version) are so central to the fairness literature; UCI Adult appeared in 300+ papers related to fairness according to Ding et al. It's something we're still discussing as a team.

Reviewer 8: That sounds very reasonable. Did you happen to already open an issue for this? That might help to get the discussion started and gather some views from the community.

Author: I can open one this afternoon :)

---------

Introduction
^^^^^^^^^^^^

The ACSIncome dataset is one of five datasets created by Ding et al. [1]_ 
as an improved alternative to the popular UCI Adult dataset [2]_.
Briefly, the UCI Adult dataset is commonly used as a benchmark dataset 
when comparing different algorithmic fairness interventions. ACSIncome offers 
a few improvements, such as providing more datapoints (1,664,500 vs. 48,842) 
and more recent data (2018 vs. 1994). Further, the binary labels in the UCI 
Adult dataset indicate whether an individual earned more than $50k US dollars 
in that year. Ding et al. show that the choice of threshold impacts the 

> amount of disparity, so they allow users to define any threshold 

Reviewer 8: Maybe mention what type of disparity you're referring to here, e.g., something like "disparity in proportion of positives".

Author: Thanks! Updated the wording

rather than fixing it at $50k.

Ding et al. compiled data from the American Community Survey (ACS) Public 
Use Microdata Sample (PUMS). Note that this is a different source than the 
Annual Social and Economic Supplement (ASEC) of the Current Population 
Survey (CPS) used to construct the original UCI Adult dataset. Ding et al. 
filtered the data such that ACSIncome only includes individuals above 16 
years old who worked at least 1 hour per week in the past year and had an 
income of at least $100.

.. _acsincome_dataset_description:

Dataset Description
^^^^^^^^^^^^^^^^^^^
Ding et al. provide data from 2014-2018 for all 50 states and Puerto Rico.
Note that Puerto Rico is the only US territory included in this dataset.

> We uploaded the 2018 data to OpenML, which anyone can access.

Reviewer 8: Nitpicks:

Maybe add a link to openml:

```suggestion
We uploaded the 2018 data to [OpenML](https://www.openml.org), which anyone can access.
```

I'm also not super sure about the phrasing of this sentence, but that's more of a style thing :)

Author: Thanks for the suggestion! I added the url for this dataset

         11. Grade 8
         12. Grade 9
         13. Grade 10
         14. Grade 11
         15. Grade 12 (no diploma)
         16. Regular high school diploma
         17. GED or alternative credential
         18. Some college but less than 1 year
         19. 1 or more years of college credit but no degree
         20. Associate's degree
         21. Bachelor's degree
         22. Master's degree
         23. Professional degree beyond a bachelor's degree
         24. Doctorate degree

   *  - MAR
      - Marital status:
         1. Married
         2. Widowed
         3. Divorced
         4. Separated
         5. Never married or under 15 years old

   *  - OCCP
      - Occupation. There are over 500 categories. Please see data dictionary at `ACS PUMS documentation <https://www.census.gov/programs-surveys/acs/microdata/documentation.2018.html>`_ for the full list of occupation codes.

   *  - POBP
      - Place of birth. There are over 200 categories, including the 50 US states and several countries. Please see the data dictionary at `ACS PUMS documentation <https://www.census.gov/programs-surveys/acs/microdata/documentation.2018.html>`_ for the full list.

   *  - RELP

>       - Relationship:

Reviewer 8: I think this feature could use a bit more explanation. Relationship to whom? The reference person?

Author: Thanks for this question. Page 41 of 2018 US Census Data Dictionary ([link](https://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMS_Data_Dictionary_2018.pdf)) and page 20 of the paper's SI ([link](https://arxiv.org/pdf/2108.04884.pdf)) only say relationship with no other information provided. However, I found this separate appendix from the US Census ([link](https://www.census.gov/content/dam/Census/library/publications/2020/acs/acs_pums_handbook_2020_appendix_a.pdf)) that says this is "relationship to householder". I'll update the description.
